qt: Handle exceptions instead of crash

hebasto opened the pull request:
Related issue #18643.

With this PR the GUI handles the wallet-related exception, displays it to a user:

![Screenshot from 2020-05-06 18-16-51](https://user-images.githubusercontent.com/32963518/81194870-d306f800-8fc5-11ea-860a-b9510bb3bd2b.png)

and, if the exception is a non-fatal error, leaves the main window running.

MarcoFalke commented:
Do I understand correctly that every gui element needs to have this handler if a crash wants to be avoided?

I.e. every dialog or window that could possibly pop up and just putting this catch in the "main gui" won't work?

MarcoFalke commented:
Concept ACK either way

hebasto commented:
> Do I understand correctly that every gui element needs to have this handler if a crash wants to be avoided?
> 
> I.e. every dialog or window that could possibly pop up and just putting this catch in the "main gui" won't work?

Every slot.

See: https://doc.qt.io/qt-5/exceptionsafety.html#signals-and-slots

MarcoFalke commented:
So currently it is undefined behaviour. Wow, qt is scary.

Slightly stronger concept ACK, but this also makes we want to remove qt.

hebasto commented:
> So currently it is undefined behaviour. Wow, qt is scary.
> 
> Slightly stronger concept ACK, but this also makes we want to remove qt.

Unrelated, but do we have a long-term plan of moving from Qt?

ryanofsky commented on src/qt/transactionview.cpp:
```diff
@@ -443,6 +444,15 @@ void TransactionView::bumpFee()
 
+void TransactionView::bumpFee()
+{
+    try {
+        bumpFeeHelper();
+    } catch (const NonFatalCheckError& e) {
+        Q_EMIT nonFatalCheckError(QString::fromStdString(e.what()));
+    }
```
No reason to tie this to to the CHECK_NONFATAL implementation or to have exceptions pass through QT code which by design does not use them. Would change this to not be tied to a specific exception type:

```
try {
    bumpFeeHelper();
} catch (const std::exception& e) {
    Q_EMIT runawayException(QString::fromStdString(e.what()));
} catch (...) {
    Q_EMIT runawayException("Unknown failure occurred");
}
```

It looks like other code could stay the same except to renaming `nonFatalCheckError` to `runawayException`

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-624710077).

ryanofsky commented on src/qt/bitcoingui.cpp:
```diff
@@ -1101,6 +1101,11 @@ void BitcoinGUI::message(const QString& title, QString message, unsigned int sty
 
+void BitcoinGUI::handleNonFatalCheckError(const QString& error)
+{
+    QMessageBox::warning(nullptr, "NonFatalCheckError exception", error);
```
Would change dialog title to something more understandable like "Internal error", since if this triggers something's gone wrong like memory corruption or a bug triggered in the code

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-624710077).

hebasto commented:
Updated 40a622dd9374e8cba80c6150f56f9d43004d953a -> 44f7f519e5fc7aef6df12474ce3c9b8583e6f20b ([pr18897.01](https://github.com/hebasto/bitcoin/commits/pr18897.01) -> [pr18897.02](https://github.com/hebasto/bitcoin/commits/pr18897.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18897.01..pr18897.02)):

- addressed @ryanofsky's comments

ryanofsky commented:
This looks pretty reasonable, and I'd ack this if not a draft PR.

One thing that might be nice is a generic exception catcher QObject that emits a handleRunawayException signal. This way the signal doesn't need to be declared multiple places and repeated boilerplate for catching exceptions could go away. So something like:

```c++
m_exception_catcher->exec([&]{ sendButtonClickedHelper(); });
```

could replace:

```c++
try {
    sendButtonClickedHelper();
} catch (const std::exception& e) {
    Q_EMIT runawayException(QString::fromStdString(e.what()));
} catch (...) {
    Q_EMIT runawayException(QString("Unknown failure occurred."));
}
```

promag commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -461,6 +462,18 @@ void SendCoinsDialog::on_sendButton_clicked()
+    try {
+        sendButtonClickedHelper();
+    } catch (const std::exception& e) {
+        Q_EMIT runawayException(QString::fromStdString(e.what()));
```
Why emit a signal? No one is interested in this signal right? Could just call another slot in a generic catcher, like:
```cpp
template <typename F> void shieldException(F&& f) {
    try {
        f();
    } catch (...) {
        QMetaObject::invokeMethod(qApp, "messageException", Qt::QueuedConnection,
            Q_ARG(QString, exception_message), ...
        );
    }
}

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-625769660).

promag reviewed (commented):
There's this approach:
```cpp
bool BitcoinApplication::notify(QObject* receiver, QEvent* event)
{
    try {
        return QApplication::notify(receiver, event);
    } catch (const std::exception& ex) {
        // ..
    } catch (...) {
        // ..
    }
    return false;
}
```
But the documentation says this will be deprecated.

> So currently it is undefined behaviour. Wow, qt is scary.
> 
> Slightly stronger concept ACK, but this also makes we want to remove qt.

Why? This is by design, Qt doesn't want to handle exceptions, what should it do anyway?

hebasto committed:
qt: Add shieldException() function

Throwing an exception from a slot invoked by Qt's signal-slot connection
mechanism is considered undefined behavior, unless it is handled within
the slot. The shieldException() function should be used for exception
handling within slots.

hebasto committed:
qt: Handle exceptions in SendCoinsDialog::on_sendButton_clicked slot

hebasto committed:
qt: Handle exceptions in TransactionView::bumpFee slot

hebasto commented:
@ryanofsky @promag Thank you for your reviews.

Updated 44f7f519e5fc7aef6df12474ce3c9b8583e6f20b -> 24a90d54ed31fd3da787b95e29e2fca581235836 ([pr18897.02](https://github.com/hebasto/bitcoin/commits/pr18897.02) -> [pr18897.03](https://github.com/hebasto/bitcoin/commits/pr18897.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18897.02..pr18897.03)).

Ready for review now :)

promag commented on src/qt/shieldexception.h:
```diff
@@ -0,0 +1,34 @@
+
+#include <cassert>
+
+template <typename FunctionType> void shieldException(FunctionType&& f) {
```
Move to src/qt/guiutil.h?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-625844423).

hebasto commented:
Updated 24a90d54ed31fd3da787b95e29e2fca581235836 -> 82529bc4ed82bcdf42428afef63c12ff37dd1dd4 ([pr18897.03](https://github.com/hebasto/bitcoin/commits/pr18897.03) -> [pr18897.04](https://github.com/hebasto/bitcoin/commits/pr18897.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18897.03..pr18897.04)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/18897#discussion_r422140790):
> Move to src/qt/guiutil.h?

hebasto commented:
Also #18435 could be related.

hebasto commented:
Rebased 82529bc4ed82bcdf42428afef63c12ff37dd1dd4 -> 1fb68b3c6cd3d153a5701b5e82f18b1f6bf96586 ([pr18897.04](https://github.com/hebasto/bitcoin/commits/pr18897.04) -> [pr18897.05](https://github.com/hebasto/bitcoin/commits/pr18897.05)) due to the conflict with #18914.

promag commented on src/qt/guiutil.h:
```diff
@@ -273,6 +279,30 @@ namespace GUIUtil
+            f();
+        } catch (const NonFatalCheckError& e) {
+            PrintExceptionContinue(&e, "Internal error");
+            ok = QMetaObject::invokeMethod(qApp, "handleNonFatalException", Qt::DirectConnection, Q_ARG(QString, QString::fromStdString(e.what())));
```
Why can't it be `AutoConnection`? That way this could be used in a "worker" slot.

Otherwise I'd assert the current thread is the GUI thread at the function begin.

hebasto replied:
I'd keep `Qt::DirectConnection` to be sure the exception is handled in the same thread where it was thrown.

As `QMessageBox` in `BitcoinApplication::handleNonFatalException()` has its own event loop, I think it is safe to call `shieldException()` from a non-GUI thread, no?

promag replied:
>  has its own event loop, I think it is safe to call shieldException() from a non-GUI thread, no?

I don't think so, only the main (GUI) thread should deal with widgets.

> I'd keep `Qt::DirectConnection` to be sure the exception is handled in the same thread where it was thrown.

What you mean by handled? This is just catching the exception (which happens in the right thread) and then showing the message. If you want to block until the user presses OK then try `GUIUtil::blockingGUIThreadConnection.`

hebasto replied:
> I don't think so, only the main (GUI) thread should deal with widgets.

You are right. [Fixed](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-627740852).

jonasschnelli commented:
Concept ACK

hebasto commented:
Updated 1fb68b3c6cd3d153a5701b5e82f18b1f6bf96586 -> 8f9424329f05ffa92a7421990c5c7320c7556a09 ([pr18897.05](https://github.com/hebasto/bitcoin/commits/pr18897.05) -> [pr18897.06](https://github.com/hebasto/bitcoin/commits/pr18897.06), [diff](https://github.com/hebasto/bitcoin/compare/pr18897.05..pr18897.06)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/18897#discussion_r424047353):
> I don't think so, only the main (GUI) thread should deal with widgets.

ryanofsky reviewed (approved):
Code review ACK 8f9424329f05ffa92a7421990c5c7320c7556a09. This looks very good and much simpler than before! 

IMO it would be nice to have a followup PR that eliminated the one-line forwarding methods and just used shieldException directly at the call and signal connect sites. Alternately renaming s/Helper/Unsafe/ would be clearer than status quo. But best to just get rid of the indirections.

fanquake commented:
Given this only had the 1 ACK (with suggested improvements), and now needs a rebase, lets move this over the [GUI repo](https://github.com/bitcoin-core/gui). I'll move #18643 over there as well.

ryanofsky commented:
> lets move this over the [GUI repo](https://github.com/bitcoin-core/gui)

Reopening this not to lose track, since it doesn't appear this PR was moved to the GUI repo (or I can't find it).  Can close this again if it is actually moved or no longer applicable.

fanquake commented:
@hebasto can you either move this over to the GUI repo, or close and we'll mark it up for grabs? The interested re-opener can open it in the gui repo.

hebasto commented:
@fanquake 
> @hebasto can you either move this over to the GUI repo, or close and we'll mark it up for grabs? The interested re-opener can open it in the gui repo.

https://github.com/bitcoin-core/gui/pull/260
